src/Main.hs

```diff
@@ -26,55 +26,49 @@ module Main where
 
 import Codec.Midi
 import Control.Monad
+import Data.Maybe
 import Data.List.Split
 import Data.Typeable
 import Prelude
 import Segno.Music
 import System.IO
 
-data GMain =
-  GMain { tokens :: [[String]] }
-  deriving (Show, Eq)
-
-tup f1 f2 f3 =
-  foldr
-    (\t trs ->
-       case t of
-         (a:[])     -> f1 a trs
-         (a:b:[])   -> f2 a b trs
-         (a:b:c:[]) -> f3 a b c trs
-         _          -> "?" : trs)
-    []
-
-one = tup (\a trs -> a : trs) (\a _ trs -> a : trs) (\a _ _ trs -> a : trs)
-
-two = tup (\_ trs -> trs) (\_ b trs -> b : trs) (\_ b _ trs -> b : trs)
-
-thr = tup (\_ trs -> trs) (\_ _ trs -> trs) (\_ _ c trs -> c : trs)
+data ChannelData = Channel1 String | Channel2 String | Channel3 String
+    deriving (Show, Eq)
 
 codecMulti n =
-  Midi {fileType = MultiTrack, timeDiv = TicksPerBeat 24, Codec.Midi.tracks = n}
+  Midi {fileType = MultiTrack, timeDiv = TicksPerBeat 24, tracks = n}
+
+dumpToFile input = do
+  let tr1 = [ a | Channel1 a <- input ]
+  let tr2 = [ a | Channel2 a <- input ]
+  let tr3 = [ a | Channel3 a <- input ]
+  putStr "Channel 1> "
+  print tr1
+  putStr "Channel 2> "
+  print tr2
+  putStr "Channel 3> "
+  print tr3
+  let m1 = tr1 >>= notes 1
+  let m2 = tr2 >>= notes 2
+  let m3 = tr3 >>= notes 3
+  exportFile "output.mid" (codecMulti [m1, m2, m3])
```

Comment:
This makes use of the fact that Lists are a Monad. If you are not comfortable with that, yet, consider using list comprehension as in lines 43-45.

Reply:
I'm fine with that, Might have to read on Monads.